File: lib/ui/tabs/playlists_tab.dart
import 'package:easy_localization/easy_localization.dart';
import 'package:flutter/material.dart';
import 'package:groovybox/data/db.dart';
import 'package:groovybox/data/playlist_repository.dart';

import 'package:groovybox/ui/screens/playlist_detail_screen.dart';
import 'package:hooks_riverpod/hooks_riverpod.dart';
import 'package:material_symbols_icons/symbols.dart';
import 'package:styled_widget/styled_widget.dart';

class PlaylistsTab extends HookConsumerWidget {
  const PlaylistsTab({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final repo = ref.watch(playlistRepositoryProvider.notifier);

    return Scaffold(
      body: Column(
        children: [
          ListTile(
            leading: const Icon(Symbols.add),
            trailing: const Icon(Symbols.chevron_right).padding(right: 8),
            title: Text(context.tr('createOne')),
            subtitle: Text(context.tr('addNewPlaylist')),
            onTap: () async {
              final nameController = TextEditingController();
              final name = await showDialog<String>(
                context: context,
                builder: (context) => AlertDialog(
                  title: Text(context.tr('newPlaylist')),
                  content: TextField(
                    controller: nameController,
                    decoration: InputDecoration(
                      labelText: context.tr('playlistName'),
                    ),
                    autofocus: true,
                  ),
                  actions: [
                    TextButton(
                      onPressed: () => Navigator.pop(context),
                      child: Text(context.tr('cancel')),
                    ),
                    TextButton(
                      onPressed: () =>
                          Navigator.pop(context, nameController.text),
                      child: Text(context.tr('create')),
                    ),
                  ],
                ),
              );
              if (name != null && name.isNotEmpty) {
                await repo.createPlaylist(name);
              }
            },
          ),
          const Divider(height: 1),
          Expanded(
            child: StreamBuilder<List<Playlist>>(
              stream: repo.watchAllPlaylists(),
              builder: (context, snapshot) {
                if (!snapshot.hasData) {
                  return const Center(child: CircularProgressIndicator());
                }
                final playlists = snapshot.data!;

                if (playlists.isEmpty) {
                  return Center(child: Text(context.tr('noPlaylistsYet')));
                }

                return ListView.builder(
                  itemCount: playlists.length,
                  itemBuilder: (context, index) {
                    final playlist = playlists[index];
                    return ListTile(
                      leading: const Icon(Symbols.queue_music),
                      title: Text(playlist.name),
                      subtitle: Text(
                        '${context.tr('createdAt')} ${playlist.createdAt.day}/${playlist.createdAt.month}/${playlist.createdAt.year}',
                      ),
                      trailing: IconButton(
                        icon: const Icon(Symbols.delete),
                        onPressed: () => repo.deletePlaylist(playlist.id),
                      ),
                      onTap: () {
                        Navigator.push(
                          context,
                          MaterialPageRoute(
                            builder: (context) =>
                                PlaylistDetailScreen(playlist: playlist),
                          ),
                        );
                      },
                    );
                  },
                );
              },
            ),
          ),
        ],
      ),
    );
  }
}

File: ui/tabs/playlists_tab.py
import flet as ft
from data import playlist_repository as prepo
from data.models import Playlist
from logic.localize import tr


def build_playlists_list(page: ft.Page, on_open: callable) -> ft.Control:
    playlists = prepo.watch_all_playlists()

    def create_new(e):
        name_f = ft.TextField(
            label=tr("playlistName"),
            autofocus=True,
            border_radius=12,
        )

        def do_create(e):
            name = name_f.value
            page.pop_dialog()
            if name and name.strip():
                if prepo.find_by_name(name.strip()):
                    page.show_dialog(ft.SnackBar(ft.Text(tr("playlistExists").replace("{}", name.strip()))))
                    return
                prepo.create_playlist(name.strip())
                page.update()

        bs = ft.BottomSheet(
            content=ft.Container(
                padding=24,
                content=ft.Column(
                    tight=True,
                    controls=[
                        ft.Text(tr("newPlaylist"), size=18, weight=ft.FontWeight.BOLD),
                        name_f,
                        ft.Row(
                            alignment=ft.MainAxisAlignment.END,
                            controls=[
                                ft.TextButton(tr("cancel"), on_click=lambda e: page.pop_dialog()),
                                ft.FilledButton(tr("create"), on_click=do_create),
                            ],
                        ),
                    ],
                ),
            ),
        )
        page.show_dialog(bs)

    tiles = [
        ft.ListTile(
            leading=ft.Icon(ft.Icons.ADD),
            trailing=ft.Icon(ft.Icons.CHEVRON_RIGHT),
            title=ft.Text(tr("createOne")),
            subtitle=ft.Text(tr("addNewPlaylist")),
            on_click=create_new,
        ),
        ft.Divider(height=1),
    ]

    if not playlists:
        tiles.append(
            ft.Container(
                expand=True,
                alignment=ft.Alignment(0, 0),
                content=ft.Text(tr("noPlaylistsYet"), color=ft.Colors.GREY),
            )
        )
    else:
        def delete_pl(pid):
            prepo.delete_playlist(pid)
            page.update()

        for pl in playlists:
            tiles.append(
                ft.ListTile(
                    leading=ft.Icon(ft.Icons.QUEUE_MUSIC),
                    title=ft.Text(pl.name),
                    subtitle=ft.Text(f"{tr('createdAt')} {pl.created_at[:10]}" if pl.created_at else ""),
                    trailing=ft.IconButton(ft.Icons.DELETE, on_click=lambda e, pid=pl.id: delete_pl(pid)),
                    on_click=lambda e, p=pl: on_open(p),
                )
            )

    return ft.Column(tight=True, controls=tiles)
